src/python/pants/option/option_types.py

```diff
@@ -0,0 +1,481 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from enum import Enum
+
+# NB: Avoid using "Optional" and use "| None" instead, as "Option" and "Optional" look similar.
+from typing import TYPE_CHECKING, Any, Dict, Generic, Type, TypeVar, cast, overload
+
+from pants.option import custom_types
+from pants.util.frozendict import FrozenDict
+
+if TYPE_CHECKING:
+    from pants.option.subsystem import Subsystem
+
+SubsystemT = TypeVar("SubsystemT", bound="Subsystem")
+OptionT = TypeVar("OptionT", bound="OptionBase")
+PropType = TypeVar("PropType")
+StrT = TypeVar("StrT", str, "str | None")
+IntT = TypeVar("IntT", int, "int | None")
+FloatT = TypeVar("FloatT", float, "float | None")
+BoolT = TypeVar("BoolT", bool, "bool | None")
+EnumT = TypeVar("EnumT", bound=Enum)
+ValueType = TypeVar("ValueType")
+# NB: We don't provide constraints, as our `XListOption` types act like a set of contraints
+ListMemberType = TypeVar("ListMemberType")
+
+
+# ================================================================
+
+
+class OptionBase(Generic[PropType]):
+    """Descriptor for subsystem options.
+
+    Clients shouldn't use this class directly, instead use one of the concrete classes below.
+
+    This class serves two purposes:
+        - Collect registration values for your option.
+        - Provide a typed property for Python usage
+
+    In order to define the `type` registration option, `self.option_type` must return a valid
+    registration type. This can either be set using a class variable or by passing it into
+    `__new__`.
+
+    NOTE: Due to https://github.com/python/mypy/issues/5146 subclasses unfortunately need to provide
+    overloaded `__new__` methods, as subclasses do not inherit overloaded function's annotations.
+    Use one of the existing subclasses as a guide on how to do this.
+    """
+
+    flag_names: tuple[str, ...]
+    kwargs: Dict
+    option_type: Any  # NB: This should be some kind of callable that returns PropType
+
+    # NB: Due to https://github.com/python/mypy/issues/5146, we try to keep the parameter list as
+    # short as possible to avoid having to repeat the param in each of the 3 `__new__` specs that
+    # each subclass must provide.
+    # If you need additional information, this class follows the "Builder" pattern. Use the
+    # "builder" methods to attach additional information. (E.g. see `advanced` method)
+    # NB: We define `__new__` rather than `__init__` because otherwise subclasses would have to
+    # define 3 `__new__`s AND `__init__`.
+    def __new__(
+        cls,
+        *flag_names: str,
+        default: Any,
+        option_type: Type | None = None,
+        help: str,
+    ):
+        self = super().__new__(cls)
+        self.flag_names = flag_names
+        if option_type is None:
+            option_type = cls.option_type
+
+        self.kwargs = dict(
```

Comment:
How about we make this private with `_kwargs` so that call sites don't abuse it and change the values? You must use the Builder methods to do that.

Reply:
I can expose a frozendict property for it and flag names (which I think is already a tuple, but still)

Reply:
Would that work, given that you mutate `self.kwargs`?

Reply:
Well the plan was to have a property of type `FrozenDict`, but that doesn't work for the dict option. let's kill it for now and come back to it :smile:

Reply:
Although killing it will alter the registration order of any subsystem which uses both styles. So let's remember not to touch those subsystems just yet.

Reply:
Oh but list options too. SO there goes `FrozenDict`. Really I just want an immutable dict but whatever no big deal.